Add auditmatica code

tw4l opened the pull request:
Connected to https://github.com/archivematica/Issues/issues/1341

This PR adds a Python `auditmatica` package that, as its primary function, generates Common Event Format (CEF) event logs from an Archivematica nginx access log. The main functions that do this work are `auditmatica.access_log.parse_access_log_line` (which parses an nginx access log line into a formatted dictionary), `auditmatica.access_log.add_event_info` (which compares the parsed line to Archivematica and Storage Service event mappings in `auditmatica.access_log.events` and annotates the parsed line with event information) and `auditmatica.cef.write_cef_event` (which writes the CEF event string).

These are in turn used by a CLI, written using Click, that provides a simple user interface: `auditmatica write-cef [options] [LOG]`. The resulting CEF logs can be written to stdout (default), a file, or syslog.

Storage Service log lines are distinguished from those of Archivematica through use of the `--ss-base-url` option.

I also put much more minimal effort into a second CLI subcommand, `auditmatica overview [options] [LOG]`. This is intended as a demonstration of how the package could be used to gain insight into the use of an Archivematica instance.

Edit: Adding that the Archivematica and Storage Service event mappings are incomplete. I tried to focus on the events that seeemed the most important from a security perspective (e.g. accessing stored data, adding or editing users, and editing configuration details). This Google doc provides some context into which events are (and are not yet) included: https://docs.google.com/document/d/1ufo9rlH7Gff9hvWcBYrxo0UDZ0wOmS9GlbD6U3Ll1AI/edit?usp=sharing. Right now that's only accessible to Artefactual staff. I'd be interested in your thoughts @ross-spencer on whether a distilled version of that document should live in this repo (maybe as a CSV?) so we can track progress moving forward.

tw4l committed:
Add auditmatica package

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+@click.option(
+    "--ss-base-url",
+    type=str,
+    default="http://127.0.0.1:62081",
```
```suggestion
    default=DEFAULT_SS_BASE_URL,
```
It'd be cool to share some of these values about. As we get into different configurations or future code changes we'll want just one place to configure these defaults/change them in code.

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+    STORAGE_SERVICE_NAME,
+    DateParseError,
+    LogParseError,
+    parse_access_log_line,
```
```suggestion
    parse_access_log_line,
    DEFAULT_SS_BASE_URL,
```

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+    "--ss-base-url",
+    type=str,
+    default="http://127.0.0.1:62081",
+    help="Base URL for Archivematica Storage Service (default='http://127.0.0.1:62081')",
```
This line might change as a result of above too.

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+    "--ss-base-url",
+    type=str,
+    default="http://127.0.0.1:62081",
+    help="Base URL for Archivematica Storage Service (default='http://127.0.0.1:62081')",
```
As above, trying to reduce duplication as much as possible should help further down the line.

ross-spencer commented on tox.ini:
```diff
@@ -0,0 +1,33 @@
+
+import-order-style = pep8
+
+[isort]
```
Interesting. I don't think isort is executed at all? I was going to bring it up. Can we add it?

tw4l replied:
Oops, I had commented it out in `.pre-commit-config.yaml` and never uncommented it! Fixed that and ran it :)

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+def overview(log, ss_base_url):
+    """Print overview of user activities from nginx access log."""
+    log_lines = log.readlines()
+
```
If we do a quick `if not log_lines:` here we can exit early and avoid any unintended behavior below. (Even though the script seems to work pretty nicely even if an empty file is piped in).

ross-spencer replied:
On that note. When we get further into the execution of the script below there isn't a lot of feedback for other scenarios like when no data is found either through configuration error or simply being supplied the wrong information. It looks like there is more scope for `stderr` logging here to provide that feedback to the user. 

A couple of examples I was thinking of going through the README were: 

* urls searched for in the log. 
* no. lines read. 
* a simple "nothing found in the logs". 
* identify where no log lines matched.
* identify when user is configured in nginx but no user in log, i.e. AM is likely to be incorrect version.

If the script is run with just a random text sample, then just that basic: "i processed something and found nothing" would be cool. But there might be nice ways to expand on that.

tw4l replied:
I added new stderr reporting in https://github.com/artefactual-labs/auditmatica/commit/76edf160932b1f4e185c4ef47b129d8dabef321d, including:
* Number of lines read in log
* "Nothing found in logs"
* A warning when no usernames are found
* A better breakdown of specific vs. default events

ross-spencer replied:
Nice! It looks really good. Thanks Tessa.

ross-spencer commented on cli/syslog.py:
```diff
@@ -0,0 +1,23 @@
+"""Syslog constants."""
```
I might accidentally duplicate this comment in the README.md. My question is really why these facilities? The rationale might sit quite nicely in the README or in this docstring. (Probably both). I don't know enough about what we're doing with syslog and why it is important, but I am looking to learn, so might suggest for that to be more explicit in the README specifically.

tw4l replied:
Good point, and thanks for the prompt to have another log at the syslog facilities. Initially I had added in all of the facilities that Python's `logging.handlers.SysLogHandler` had support for, to retain a certain amount of flexibility (and because I also don't know a ton about syslog, other than what I've read since noticing that the CEF specification puts emphasis on using it when possible).

Having looked again, I don't think it's appropriate for `auditmatica` to write to many of those facilities, so I have kept the default of `USER` and restricted the `--syslog-facility` options to only `LOCAL0-LOCAL7`, which are facilities whose purpose is left open for applications to use. This is closer to what other applications do, e.g. [Postgres](https://postgresqlco.nf/doc/en/param/syslog_facility/).

Two sources I found helpful on the subject.
- https://en.wikipedia.org/wiki/Syslog#Facility
- https://unix.stackexchange.com/questions/90842/what-is-the-local6-and-all-other-local-facilities-in-syslog

I'll update the `cli/syslog.py` module docstring and README accordingly.

ross-spencer commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+    url="https://github.com/artefactual-labs/auditmatica",
+    author="Artefactual Systems",
+    author_email="info@artefactual.com",
+    license="AGPL 3",
```
Is this line in conflict with line 24? MIT license?

tw4l replied:
Yes, it is! That's a copy-paste error on my part. It should be AGPL 3 on line 24.

ross-spencer commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    url="https://github.com/artefactual-labs/auditmatica",
+    author="Artefactual Systems",
```
Did we lose the `inc`? (I might be the only one that ever references it!)

tw4l replied:
Nope, I just wasn't being consistent! Fixed it

ross-spencer commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+from setuptools import setup, find_packages
```
PIP for this tool works really nicely. It's quite a simple but effective setup here. Nice :+1:

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+            if parsed_line:
+                parsed_log_lines.append(parsed_line)
+        except (DateParseError, LogParseError) as err:
+            errors.append(err)
```
Does anything happen to these errors after this point?

tw4l replied:
They are now included in the JSON output if present

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+        click.echo("\n")
+        click.echo("Archivematica users per event:")
+        for event_name in am_events_by_name:
+            _print_event_users(parsed_log_lines, event_name)
```
I feel like a builder like pattern is going to benefit code readability and maintainability across these loops and the ones below for the storage service. 

Where I think we need to get to is:

* Construct the data structure with all of our data. 
* Output it once. 

Whereas, right now, we're making some data, formatting it, and outputting it. Then repeating. The `_print_event_users` function caught me off guard a little by not returning something to print here so that all of our representation output is in one place.

One suggestion to keep things as they look now is to build a string. Concatenate as you go and have `_print_event_users` return a string value that can then be concatenated to. And then add the formatting as is being done. 

You may have considered this, but it might be good to ask what you think of a standard data structure here? YAML looks closest to your format. JSON is the one I know we get a lot free out of the box. 

An example of something I tried locally was:

```
    output["Archivematica users per event"] = am_events_by_name
    output["Archivematica users by count"] = am_events_by_count
    output[storage service etc. etc.) = ... 
    click.echo(json.dumps(output, indent=4, sort_keys=True))
```

The `json.dumps` provides user-ish friendly formatting. And of course it's only output to screen once. But it can then be consumed using other standard tooling too. 

What do you think about trying something like that? A good-scenario would be to have a single write to screen - keeping logic as separated as possible from presentation. I'd also be really interested to see what can be done with 
 `dict` or somesuch. 

Two problems I didn't explore in too much detail is ordering the output which I can see you're working really hard to make a feature. And then the structure I wrote is just a dict of dicts. You'd probably want an array in there to order it `{[am-events, am-events-by-user, ss-events, etc.]}`.

tw4l replied:
Fixed in https://github.com/artefactual-labs/auditmatica/commit/76edf160932b1f4e185c4ef47b129d8dabef321d, which now constructs the data and then outputs it all at once as JSON. Nice guidance here!

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+def _print_event_users(parsed_log_lines, event_name):
+    """Print comma-separated list of users for specified event."""
+    event_users = set(
+        line.get("username")
```
Is there a sensible alternative default here for `None`, e.g. `No user specified`. From my understanding this could arise from having nginx configured but Archivematica either on the wrong version or not configured to output this header. By the time we get the complete overview report it would be good to have something in the field rather than a blank space (that's how I see it with some faulty logs on my end anyway).

tw4l replied:
The `overview` command now defaults to "Unauthenticated user" as of commit https://github.com/artefactual-labs/auditmatica/commit/76edf160932b1f4e185c4ef47b129d8dabef321d when appropriate.

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+        click.echo(f"Errors: {len(errors)}", err=True)
+        if verbose:
+            for error in errors:
+                click.echo(error, err=True)
```
```suggestion
                click.echo(str(error).rstrip("\n), err=True)
```
I believe these strings arriving here will already be newline terminated and so we can save space by stripping that. Echo will otherwise add another.

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+from auditmatica.events import DEFAULT_EVENT, EVENT_SIGNATURE
+from cli.syslog import SYSLOG_FACILITY
+
+logger = logging.getLogger()
```
The pieces around logging in this script seem really clever. Nice work. 

For clarity, are we able to change the name of the global logger to something like `cef_logger` that for me would make it clearer that we're not outputting some other local script base logging, while also leaving open the potential to write local logging if we ever want to write out to stderr using logger... if that makes sense! `cef_logger` or similar would be awesome if possible.

tw4l replied:
That makes a lot of sense! Good suggestion :)

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+    consisting of date and host, and begin instead with "CEF:0|...".
+
+    Datestamps from the nginx acess logs are included separately in the
+    "rt" (requestTime) field in the CEF extensions.
```
It has been a long day looking at code I can't see in this function if we're doing this? e.g. Are we using this function to remove the syslog prefix? If not, I might suggest removing 36-41. My reading thus far is the formatting is done out of here and we just output logging as it arrives at the logger.

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+        # or a string (e.g. "/dev/log"). Here we default to the string
+        # format but use a tuple instead if a port is provided.
+        if syslog_port:
+            syslog_address = (syslog_address, syslog_port)
```
What happens if I run this but I accidentally provide a port, but the address as `/dev/log`?

tw4l replied:
It now prints a warning and ignores the value `--syslog-port` if an address isn't provided or is `/dev/log`

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,258 @@
+@click.group()
+def auditmatica():
+    """Auditmatica: Archivematica auditing package"""
+    pass
```
What does this do?

tw4l replied:
This [Click group](https://click.palletsprojects.com/en/7.x/api/#click.Group) enables the use of subcommands.

Each of the subcommands then begins with a decorator like so: `@auditmatica.command("write-cef")`, which links them together for Click and facilitates the automated `--help` responses. If this function were named `main`, those decorators would become `@main.command("write-cef")`.

The docstring for this function is also passed along to the `--help` documentation:

```
Usage: auditmatica [OPTIONS] COMMAND [ARGS]...

  Auditmatica: Archivematica auditing package

Options:
  --help  Show this message and exit.

Commands:
  overview   Print overview of user activities from nginx access log.
  write-cef  Write Common Event Format (CEF) log from nginx access log.

```

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,714 @@
+    # TODO: Ideally we would record the decision made for a package
+    # deletion request (approve or deny). This would require inspecting
+    # the body of the POST request, which isn't currently possible via
+    # the nginx access logs.
```
We should err away from TODOs in code and try to either do them (which reading this doesn't seem possible) or trust that one day it will be picked up on merit, or through the careful reading and understanding of other devs in the future. It can be logged elsewhere - suggest the issues. It starts to hit home when you search old code-bases and still find them and they haven't been done after years. They're also locked away because no one was looking at the code, and only the devs look at the code.

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,714 @@
+in a slash prior to attempting to make a match.
+"""
+
+HTTP_STATUS_OK = "200"
+HTTP_STATUS_CREATED = "201"
+HTTP_STATUS_ACCEPTED = "202"
+HTTP_STATUS_NO_CONTENT = "204"
+HTTP_STATUS_MOVED_PERMANENTLY = "301"
+HTTP_STATUS_FOUND = "302"
+HTTP_STATUS_UNAUTHORIZED = "401"
+HTTP_STATUS_NOT_FOUND = "404"
```
Since we're using Python 3 it might be worth considering reusing these statuses from the standard library: https://docs.python.org/3/library/http.html#http.HTTPStatus - I had a look for verb constants too but they don't seem to exist unfortunately.

tw4l replied:
TIL. Thanks Ross - these are good to know about!

After a bit of trying to use the values in `http.HTTPStatus`, I'm tempted to leave these as-is for now. The `http.HTTPStatus` status codes are enums and their values are ints, so the the equivalent for `HTTP_STATUS_OK` that lets the tests pass is `str(HTTPStatus.OK.value)`. That seems like a bit much to repeat through the code, and I'm not sure that just changing the definition of these consts to, e.g. `HTTP_STATUS_OK = str(HTTPStatus.OK.value)` quite gets the intended value of this change.

ross-spencer replied:
That seems fair!

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,714 @@
+Actual URLs from Archivematica may or may not terminate in a slash.
+To aid in comparison between actual URLs and our event mappings,
+auditmatica.access_log.add_event_info() ensures all URLs also terminate
+in a slash prior to attempting to make a match.
```
This is interesting information but I don't believe it's this script doing the work described? Perhaps it is better placed in the function docstring where it does happen? (or just removed from here if it is already there) What do you think?

ross-spencer replied:
I am wondering if this is meant to be a directive? _"Any URL recorded in the mappings below must be terminated with a forward-slash"_?

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,714 @@
+}
+
+Placeholders are replaced prior to comparison. Placeholders currently
+implemented include:
```
What comparison? Is there more explanation in this module docstring missing? Or does some of this docstring belong elsewhere? 

This module looks like it is good for describing these structures, it can probably be kept pretty simple with the functionality described in the various calls outt to this module.

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,714 @@
+    EVENT_NAME: "Attempt to access unauthorized content",
+    EVENT_SIGNATURE: 3,
+    EVENT_SEVERITY: 5,
+}
```
Suggest adopting a little-endian approach here and placing these three structures toward the top of the file as they're the shortest and easiest to understand. That would mean the AM and storage events to come afterwards.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+    :param url: URL string
+
+    :returns: Modified URL string
+    """
```
I wonder if we're doing something else here that can be more descriptive? 

It sounds like we're normalizing a URL so that it can be compared. Removing all variable data? Is that correct? 

If so I'd consider renaming the function to something like `_normalize_url_for_comparison(...)`  or similar. 

Perhaps consider another word for placeholder. Are we actually masking the string for example? The reason I ask, and am not convinced about placeholder is at this point in the code, the placeholder doesn't hold any place for anything we are going to do after? - We're removing information that we're neither going to retreive or fill-in again in future? 

I hope my description helps you where my mind is, perhaps that provide some clues. Obviously let me know if this is way off.

tw4l replied:
I see where you're getting at but all the alternatives I could think of were harder to understand than "placeholder" so I'd prefer to keep it. I did rename the function to be more descriptive.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+
+    :param url: URL string
+
+    :returns: Modified URL string
```
based on "if url" below; 

* or `None` or an empty string if an empty string or `None` was sent to this function? What are the use cases for that happening? Do we want to normalize to one or the other if not url?

tw4l replied:
Addressed by making these formatting functions less defensive, per https://github.com/artefactual-labs/auditmatica/pull/1#discussion_r573772086

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+PLACEHOLDER_PROCESSING_CONFIG = "<processing_config>"
+PLACEHOLDER_USER_PK = "<user_pk>"
+
+URLS_TO_IGNORE = [
```
Nice list!

tw4l replied:
Thanks! :smile:

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+)
+LINE_FORMAT_WITH_USERNAME = re.compile(
+    r"""(?P<ip_address>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}) - (?P<remote_user>.+) \[(?P<datetime>\d{2}\/[a-z]{3}\/\d{4}:\d{2}:\d{2}:\d{2} (\+|\-)\d{4})\] ((\"(?P<method>.+) )(?P<url>.+)(http\/[1-2]\.[0-9]")) (?P<status_code>\d{3}) (?P<bytes_sent>\d+) (["](?P<referrer>(\-)|(.+))["]) (["](?P<user_agent>.+)["]) (["](?P<forwarded_for>.+)["]) user=(?P<username>.+)""",
+    re.IGNORECASE,
```
Wild regex! Very cool. 

Do I understand correctly that this regex maps directly to the log_format we're asking users to condigure? 

```
log_format main '$remote_addr - $remote_user [$time_local] "$request" '
	'$status $body_bytes_sent "$http_referer" '
	'"$http_user_agent" "$http_x_forwarded_for" '
	'user=$upstream_http_x_username';
```

I think, if so, then I might elaborate in the README that this configuration option maps directly to a pattern that will be matched by auditmatica to identify lines to be processed to create audit logging. 

Additionally, I'dd comment these consts to describe that too, e.g. these regex patterns map directly to the string configured by the user in the `log_format` in nginx.conf. 

This helps me first as a reader, but it makes it clear the `log_format` must be exact in the README and might provide useful information that can be used to debug in the future without necessarily being a dev.

Because of the complexity of regex, I wonder too if it is worth breaking a comment down into capture groups that describe any patterns that are mandatory/optional/can be matched multiple times etc. (As devs we can always work it out, but it's so much nicer to have some of the intentions/expectations laid out up front)

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+
+from auditmatica import events
+
+ARCHIVEMATICA_NAME = "Archivematica"
```
Is this repeated in `cef.py`? I believe `cef.py` can access this - is there a risk of a developer modifying it in some other scope and impacting the behavior of the other? (unlikely, but stranger things have happend). Perhaps we can just use one or the other?

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+        event_info = events.DEFAULT_EVENT_404
+
+    if not event_info:
+        event_info = events.DEFAULT_EVENT
```
This function works well - it has a single responsibility. But it's also doing at least two things. This block in the middle requires a little reading to understand. What I see it doing is:

1. Identify and add product to our parsed log dict. (4 lines)
2. Return an event_info block and add it to the parsed log dict (17 lines).

This could be cleaned up a little to two separate internal functions doing this within here. Then too the code could describe itself a little better with the function names for both of those describing what they're doing. 

There might be other areas in the code I am seeing this pattern of a single function being resonsible for a little more work than maybe is needed but this seemed like a nice example to highlight and see if there was a way to split it out a bit more. I am also seeing a lot of good examples of code where it's obvious that this type of split is being consciously made and it's looking good!

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+    :param ss_base_url: Base URL for Storage Service, as present in
+        nginx access log "referrer" field (str)
+
+    :returns: Parsed log line with added event information (dict)
```
I am seeing `parsed_line` come out of a few of these functions. Is there a point where we're looking at more than a parsed line? `augmented_log`?

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+
+    :returns: Parsed log line with formatted URL (dict)
+    """
+    url = parsed_line.get("url")
```
See comment below about datetime.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+    :returns: Parsed log line with formatted datetime (dict)
+    """
+    nginx_datetime_str = parsed_line.get("datetime")
+    if nginx_datetime_str:
```
This check feels a little defensive at this point in the code. Based on the groupdict returned from the regex matching we will always have a datetime so we really just need to retrieve it from the dict and modify it. 

I think where my mind goes reading the function and this check is "okay, so we're going to format the datetime if we have it - but why don't we care at all if we don't have a datetime string - is there an explicit response for that? Maybe an attribute error is as good as any?" 

Perhaps that leads to two strategies - describing why we don't care in the docstring, so returns might be:

* `:returns: Parsed log line with formatted datetime (dict) or unformatted parsed log line with no date`

Or I suspect we have enough testing to know that we're okay and we can remove the check and the docstring remains accurate. If an error in the wild is preferable then using `['datatime']` as the lookup and returning an attribute error - or capturing it as a DateParseError might work quite well too.

tw4l replied:
Good point! Agreed that it's best to remove the defensive check here and in a few of the other functions in this module, as in that case we'll raise a `LogParseError` well before these functions get called anyway.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+    parsed_line = _format_datetime_string(parsed_line)
+    parsed_line = _format_none_values(parsed_line)
+    parsed_line = _ensure_username_key(parsed_line)
+    return parsed_line
```
Nice. I like how this function separates each of the responsibilities and builds its result before returning.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+
+    :returns: Flag indicating whether to filter line (bool)
+    """
+    url = parsed_line.get("url")
```
Will we ever not have a url at this point?

tw4l replied:
Nope! Defensive check removed

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+    """
+    username = parsed_line.get("username")
+    if username is None:
+        parsed_line["username"] = None
```
Do we definitely want this to be `None`. This will output as whitespace I believe in the overview report. Is there a better placeholder to consider? "no user"/"user logging not configured"/"user not specified"?

tw4l replied:
It's a good question. I'm not sure. In terms of the data in the dictionary itself, `None` seems like an accurate representation (meaning either that the user was unauthenticated, or that user logging isn't configured). I hear you about the overview though. Maybe this is something to patch up in the CLI code for that subcommand?

ross-spencer replied:
This makes more sense in JSON now with there being an empty array in the output for user.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,329 @@
+    :param ss_base_url: Base URL for Storage Service, as present in
+        nginx access log "referrer" field (str)
+
+    :returns: Parsed and formatted log line (dict)
```
I don't know what the 
```suggestion
    :returns: Parsed and formatted log line (dict) or empty dictionary if line cannot be parsed dict).
```
I'm not sure the protocol on the error being raised. Do docstrings sometimes have a `raises` component?

tw4l replied:
It seems them do! I've added them where appropriate, e.g.:
```
:returns: Parsed and formatted log line (dict)
:raises LogParseError: if log line is in unexpected format
```

ross-spencer replied:
That's great. Thanks Tessa.

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,714 @@
+EVENT_SIGNATURE = "event_signature"
+EVENT_SEVERITY = "event_severity"
+
+ARCHIVEMATICA_EVENT_MAPPING = {
```
This work all looks really good. 

The only thing in the back of my mind is that it's a `dict` which is easily modified. As you know when it is assigned in the other modules a new copy isn't made in memory, it's this one specifically so if it were accidentally modified then it would change globally. 

There's two easy ways around that. When it is assigned, one could do `a = dict(ARCHIVEMATICA_EVENT_MAPPING)` and a new copy would be created. 

Alternatively, a small API could be wrapped around some of this work in this module to do the same. 

```python
def get_am_event_mapping():
   return dict(__ARCHIVEMATICA_EVENT_MAPPING)
```
^^^ the double underscore should make mark ARCHIVEMATICA_EVENT_MAPPING as private. 

These obviously double the memory usage but might make the code a little safer. Even though we know we don't want to modify these dicts it's not easy to predict future behavior.

There are probably more complex ways of providing some protection here, but that might require changing the data structure which I don't think we should do. `metsrw` uses a complex pattern of tuples - I believe for similar reasons? (I am not too sure). 

What do you think are some options here?

tw4l replied:
Hm, this is a good point! I think I'm leaning toward your suggestion of a small API wrapper. The memory footprint of `ARCHIVEMATICA_EVENT_MAPPING` seems to be only about 2 MB, so I'm not too concerned about the increase there, and I find that easier to work with than tuples à la `metsrw`. But I will give it some more thought!

ross-spencer commented on auditmatica/cef.py:
```diff
@@ -0,0 +1,149 @@
+    replaced_uuid = parsed_line.get("replaced_uuid")
+    if replaced_uuid:
+        msg = f"UUID:{replaced_uuid}"
+    cef_message += _format_cef_extensions({"suser": username, "msg": msg})
```
I'm not used to the f-string formatting method you're using. `+=` doesn't seem right though - at least it's a syntax I hoped we had got away from! (+= equates to arithmetic in my mind) Is there another concatenation method that the formatter specifies? 

Maybe:

```
   cef_message  = "%s%s" % (cef_message, format_cef_extensions(...))`
```

Same for the use of `+=` in `_format_cef_extensions(...)`.

tw4l replied:
For consistency, the f-string approach would be something like:

```python
cef_message = f"{cef_message}{_format_cef_extensions({"suser": username, "msg": msg})}"
```

Or less concise but more readable (at least to my eyes):

```python
extensions = _format_cef_extensions({"suser": username, "msg": msg})
cef_message = f"{cef_message}{extensions}"
```

What do you think of that? I really like f-strings for their combination of brevity and self-description, but it's true that Python gives us a confusingly wide range of ways to format strings!

ross-spencer replied:
I reckon: `cef_message = f"{cef_message}{_format_cef_extensions({"suser": username, "msg": msg})}"` works well for me!

ross-spencer commented on tests/test_access_log_parser.py:
```diff
@@ -0,0 +1,1132 @@
+                "referrer": "http://127.0.0.1:62080/backlog/",
+                "user_agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:84.0) Gecko/20100101 Firefox/84.0",
+                "forwarded_for": None,
+                "username": "test",
```
I reckon at some point we'll enable the use of diacritics in user names in AM and the storage service. I wonder if we can add some diacritics here and in the fixture to get ahead of that? I am sure it will just work, but it's great to have the knowledge encoded in the tests for when that happens so the downstream change doesn't affect us here.

tw4l replied:
Absolutely, and great suggestion!

ross-spencer commented on tests/test_access_log_parser.py:
```diff
@@ -0,0 +1,1132 @@
+        # dictionary.
+        (os.path.join(FIXTURES_PATH, "access_log_with_urls_to_filter.log"), {}),
+        # TODO: Add tests for filtering /login/ URLs from Archivematica
+        # but not Storage Service.
```
As per another TODO comment I feel this needs to either be done, or committed to the GH issues for the repo. It seems important enough to just finish now? is it a big task?

tw4l replied:
I think I intended to do this before submitting for code review and just missed it! Agreed it'd be both good and feasible to add these tests in and remove the comment.

ross-spencer commented on tests/test_access_log_parser.py:
```diff
@@ -0,0 +1,1132 @@
+            },
+        ),
+        # Sample log line with URL to filter out should return empty
+        # dictionary.
```
This might be quite a broad question, i.e. broader than the tests - the term filter implies that we consciously add some criteria to something like a search pattern that narrows the number of results returned. Is that really what's happening in auditmatica? Are we simply just not seeing a matching string? one feels active, and one feels passive. That certainly impacts how I read the code and its expectations, but I am not necessarily sure of my understanding/conclusions on that. Does it make sense?

tw4l replied:
It's an interesting question! The way I have [`auditmatica.audit_log._filter_log_line_on_url`](https://github.com/artefactual-labs/auditmatica/blob/386b75c1287746b00e586584ce61b4e2bbd9e122/auditmatica/access_log.py#L182-L206) was inspired by filter functions in the functional programming sense, which is why it returns `True` if the url is one we want to ignore (e.g. is in `URLS_TO_IGNORE` or doesn't have a referrer). But it's not really a filter function per se - it's not taking a structure like a list and only returning items that match a certain set of criteria. And we'd typically use a list comprehension for that in Python anyway.

So there might be a better term for it. But I do think what's happening is an active and not passive process - lines are filtered in `auditmatica` not because they don't match the event mappings but because they match specific criteria - which makes me think filtering as a term might still be an okay fit.

tw4l replied:
I've renamed this from `_filter_log_line_on_url` to `_discard_log_line`, which returns a boolean. On further reflection I think you're right and that this is more accurate.

ross-spencer replied:
That works well. I was reviewing the code before the test and it made good sense. Thanks Tessa.

ross-spencer commented on tests/test_access_log_parser.py:
```diff
@@ -0,0 +1,1132 @@
+        # Log lines that don't match mapping should return the default
+        # event information.
+        (
+            {"status_code": "418"},  # I'm a teapot, short and stout.
```
Nice inclusion. Thanks for the giggle!

ross-spencer commented on tests/test_access_log_parser.py:
```diff
@@ -0,0 +1,1132 @@
+    assert return_value[EVENT_NAME] == expected_name
+    assert return_value[EVENT_SIGNATURE] == expected_signature
+    assert return_value[EVENT_SEVERITY] == expected_severity
+    assert return_value["product"] == ARCHIVEMATICA_NAME
```
I won't put the magnifying glass on these tests and the ones below. They look really solid and obviously very detailed. We'll clearly benefit from them. Nice work on this and the event mapping itself.

tw4l replied:
Thanks! Just in the interest of sharing experiences, I took a bit of a TDD approach with these, writing the test for each event before adding it to the event mappings. I did the same for the CEF string escaping. It made for a really nice experience, and helped me catch a couple typos/errors in thinking along the way. I'll definitely be doing more of that moving forward!

ross-spencer replied:
Nice! I can imagine that approach making quite a difference. Thanks for sharing!

ross-spencer commented on tests/test_cef.py:
```diff
@@ -0,0 +1,157 @@
+import pytest
```
All of this module looks like really solid work. Nice :+1:

ross-spencer commented on tests/test_cli.py:
```diff
@@ -0,0 +1,88 @@
+
+def string_sha256_hash(string):
+    """Return SHA256 hash for string."""
+    return hashlib.sha256(string).hexdigest()
```
Wunderbar! What do digipres peeps bring to software development? Checksums! (Maybe, a little hyperbolic but I really dig this approach - nice work again :+1: )

tw4l replied:
Thanks! This did feel like a "past and current careers coming together" moment! :smile:

ross-spencer commented on tests/test_cli.py:
```diff
@@ -0,0 +1,88 @@
+import hashlib
```
I was initially wondering where you were going with the integration tests when talking in slack but these all look really great. Super!

ross-spencer reviewed (commented):
This is good work Tessa. Exceptional effort to publish what is a very complete application. I also imagine this was a very rewarding task. The attention to detail reverse engineering the different signals presented in the nginx logs and then mapping them in the application looks incredible. It's very cool to see this effort. 

For something like this, I treat the README as an entry point to the review. This is one of the few opportunities to do this with completely new eyes so I hope the feedback is useful. I then moved from the README to the code, and I see a good alignment between the two. 

I have more specific CR notes inline as per usual. Because of the size of the PR I have more general comments below too. 

Some of my in-line notes may seem more terse than usual - apologies if so. Trying to capture the context and provide constructive feedback is a little tougher than usual again with the PR size - but I hope I've managed to help celebrate some of the great pieces of work in here. 

(NB. You asked explicitly about things that aren't implemented - I've suggested issues. For lists of things like urls not mapped, you could use a checklist style? https://github.com/archivematica/Issues/issues/1152 - i see Issues are not turned on. Is this a wider strategy in the company now? - I don't think that's a problem - archivematica/issues is fine but maybe there needs to be some labelling to clarify when it's auditmatica (an auditmatica label?) - I'm not sure how folks want to deal with that. I'd err away from a CSV as not all folks have Google and I feel there has to be a workflow in Github that works for us - could be discussions?)  (some things to ponder anyway). 

# General comments

##  Good stuff right off the bat

* CI setup immediately. 
* pip -e flag description is a nice addition.
* Code layout looks good.
* Lots of effort.
* Good use of constants in-scope/globally as required.
* First time I've seen return annotations used in our code. Seems cool!
* Cool that you've used type hinting in places too. I'd have been really interested to see the outcome if it is used more globally, is there a rationale for not using it more widely? 

## Things to think about

* Publishing scripts to enable the script to be pushed to PyPi: a good references is the mets-reader-writer repo: https://github.com/artefactual-labs/mets-reader-writer/blob/master/Makefile

* Is there a preferred default argument that should be used without requiring a flag? I like the idea of that for just getting going - cat <some_log> | auditmatica > output_happens_here

* Output formatting: The custom format is interesting but I am wondering if JSON/YAML would help simplify expectations and increase future uses some more? I have left some notes in the code about theat. We might get a lot for free with JSON. 

## Notes about the README

* Have a look at the style guide for docs: https://github.com/artefactual/archivematica-docs/blob/12da09e3664025169450ac8b913639ff85ebba84/style_guide.md particularly around line-length. We're expecting 80 in most cases and our repositories mostly should use this standard on the README now. 

* Additionally, you might find some value in running markdownlint afterwards. We only use it in the Architectural Decision Records repo at the moment so it's not a true standard we adopt but I usually find value in it minting new markdown. I ran it locally and it seems like there might be some benefits to the ``` markup where [a language code](https://www.rubycoloredglasses.com/2013/04/languages-supported-by-github-flavored-markdown/) can be specified. This provides syntax highlighting which can make it easier for some users. 

* My instinct is to reorganize the README in order of most likely to be referenced information - so purpose/description/outputs first - we'll always want to refer back to this first. Installation/development last - ideally we only install once, or get the hang of it pretty quickly by hand/automate it afterwards. It's interesting, but not as interesting as the subject matter :wink: -- this is really just my gut feeling. I asked Peter to reorganize AIPscan that way, I reorganized Fixity that way, and try to do so on my own projects. 

Suggested: 

- About
- Overview
- Usage
- Download 
- Install
- Developer notes, e.g. PIP, Tests.

* proxy_hide_header optionally (but probably a good idea) is nice, friendly. But I'd consider making it more terse just to there's no doubt. Reading the context a little more it sounds like it's more preferrable than optional? (FWIW from reading I seem to understand the user being sent back to the client as not necessarily being bad - but I'm glad to see the option mentioned explicitly. I can imagine some asking how to remove it.))
  
* Some more information about where to access conf and log files would be nice, I'm on CentOS looking at /etc/nginx and /var/log/nginx for both respectively. I recognize this can get to be a bit of a pain with lots of ways of deploying - is there an nginx resource which describes it? (Might not cover Docker I guess). 
  
* Do users need a restart for the logging to work as expected? (we also spotted some other additions to add to the log file line). 

* CEF: More background on what it is and why we're doing it will help. Reading through the client spec I see this was their requirement. I think some narrative that balances out what it means to the client, as well as a superuser like Santi, vs. a devops noob like myself would be good. 
  
* You have a really nice section at the bottom that is pre-formatted text but looks like part-markdown: https://github.com/artefactual-labs/auditmatica/blob/386b75c1287746b00e586584ce61b4e2bbd9e122/README.md#implemented-cef-events
  
* I wonder if it is worth adding a new document SPECIFICATION.MD? IMPLEMENTATION.MD? This could all be put in that section. Then I think what would also really benefit the CEF description is linking to our specification of CEF events. And also the separate list of CEF events which a user can expect to find. Immediately as I start to understand the tool more I'm thinking how do I find a canonical list of events I can grep/search for. Severity looks cool too. This would have value as a separate document or promoted higher up in this README. Arguably then above the write-cef option for the app.   

* Python 3! Is it worth stating explicitly that we are using it?

* Have a think about splitting commands onto different lines, e.g. 

    `auditmatica write-cef -s --syslog --syslog-address localhost --syslog-port 514 --syslog-facility LOCAL0 /path/to/nginx/access.log` 
    
Becomes:
  
  ```bash
    auditmatica write-cef -s --syslog \
     --syslog-address localhost \
     --syslog-port 514 \
     --syslog-facility LOCAL0 /path/to/nginx/access.log
  ```

This should just help scanning by eye to understand what's going on.

* Consider the lanaugage around: --ss-base-url -- What I understand from the application and what I see is that there is always a differentiation between storage service and archivematica if log entries exist for both in the same log but Auditmatica looks for the docker default and if it isn't there it won't output storage service statistics. So, it's not so much a flag (which I feel the text almost suggests) as a way of overriding the default. -- I think I'd simplify the language as much as possible and just keep it at: override the storage service URL to scan for. Or something like that. That will hold true whichever log is being inspected that includes storage service entries.

* Wishlist as GH issues: But have a think about describing the problem, e.g. why gzipped logs? (I don't know!) - the unmapped events could be a simple issue with a checklist listing each. 

* For the syslog facilities, is it better to link out to another source than we listing them? The superusers looking at this page will know what to do, and we don't needto maintain that information? e.g. https://en.wikipedia.org/w/index.php?title=Syslog&oldid=1002309663#Facility - or are not all facilities valid? Is there a reason that needs to be stated in the README for that?

tw4l commented:
Hi @ross-spencer  - I believe this is now ready for re-review! I have addressed your inline comments, reworked the README, added an `IMPLEMENTATION.md` (nice suggestion!), and added a `Makefile` with commands to help publish the package to PyPI. Among the changes, the `overview` command now outputs JSON, which I'm really enjoying.

A few specific comments I wanted to address:

> Is there a preferred default argument that should be used without requiring a flag? I like the idea of that for just getting going - cat <some_log> | auditmatica > output_happens_here

I see what you mean, but the downside would be that running just `auditmatica` would no longer display the help menu, with the list of available subcommands. So I'd rather leave it as-is if you're okay with it.

> Cool that you've used type hinting in places too. I'd have been really interested to see the outcome if it is used more globally, is there a rationale for not using it more widely?

We talked about this a bit offline, but the inconsistent usage is because Python type expects you to specify the types of keys and values for `dicts`, which would require a bit of an overhaul at this stage that I don't think would bring additional value besides making the type annotations happy. So I used type annotations for the simpler types like `str` and `bool` but not for `dict`s.

tw4l commented:
One more note - I'd like to publish this to PyPI but am just waiting for code review to be complete before doing so :)

ross-spencer commented on .pre-commit-config.yaml:
```diff
@@ -10,8 +10,8 @@ repos:
-#   hooks:
-#   - id: isort
-#     language_version: python3
+- repo: https://github.com/pycqa/isort
```
Works perfectly! :+1:

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,715 @@
+
+
+def get_ss_event_mapping():
+    return dict(__STORAGE_SERVICE_EVENT_MAPPING)
```
Was there a particular rationale for leaving out a docstring for the two functions we're adding?

While these are simple functions it seems reasonable to just describe their purpose for accessing the mappings here; either in the function or module docstring. Have a think though.

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,362 @@
+# in the nginx.conf file, and are used to parse the nginx access log
+# lines into a re.Match.groupdict dictionary keyed by subgroup name.
+# Subgroups are mandatory and non-repeatable. Empty fields in the nginx
+# logs are indicated by a single dash (-) and parsed as-is.
```
Nice description :+1:

ross-spencer commented on auditmatica/access_log.py:
```diff
@@ -0,0 +1,362 @@
+def _save_uuid(augmented_line):
+    """Add UUID that will be replaced by placeholder to dict.
+
+    :param parsed_line: Augmented log line (dict)
```
It looks like there might be some shift in the docstrings following he code-changes. `parsed_line` -> `augmented_line` - it looks like there might be one or two below.

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,299 @@
+    cef_logger.setLevel(logging.INFO)
+
+
+def _process_log(log_lines: list, ss_base_url: str, suppress_unspecified: bool):
```
Is `ss_base_url` used in the function at all?

tw4l replied:
It's not, but it should be the second argument to `parse_access_log_line()`!

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,299 @@
+@click.version_option()
+def auditmatica():
+    """Auditmatica: Archivematica auditing package"""
+    pass
```
My comment may have led to the introduction of this change? You can either `pass` or `"""my function comment"""` will act like `pass` and one or the other isn't needed.

tw4l replied:
I didn't realize the docstring was sufficient on its own! I'll remove the `pass`

ross-spencer commented on cli/cli.py:
```diff
@@ -0,0 +1,299 @@
+            "error_messages": [str(error).rstrip("\n") for error in errors],
+        }
+
+    click.echo(json.dumps(output, indent=2, sort_keys=True))
```
Perfect! :ok_hand:

ross-spencer commented on tests/test_access_log_parser.py:
```diff
@@ -0,0 +1,1314 @@
+                "referrer": "http://127.0.0.1:62080/backlog/",
+                "user_agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:84.0) Gecko/20100101 Firefox/84.0",
+                "forwarded_for": None,
+                "username": "zoë",
```
Nice :+1:

ross-spencer commented on tests/test_cli.py:
```diff
@@ -0,0 +1,116 @@
+  ],
+  "Storage Service events": [
+    {
+      "count": 1,
```
One thought is that if sorting is a feature of this output, e.g. the most common events first then it would be good to test those explicitly here. So a test that just has a higher count that is expected first followed by a lower count. What do you think?

ross-spencer commented on auditmatica/events.py:
```diff
@@ -0,0 +1,715 @@
+}
+
+Placeholders are used in the mappings in place of several types of
+dynamic data that overwise complicates comparison between actual URLs
```
```suggestion
dynamic data that otherwise complicates comparison between actual URLs
```

ross-spencer reviewed (approved):
Two or three small things left Tessa but looking good:

* Docstring shifts with parameter changes in some functions. 
* One unused parameter in the cli? 
* A potential improvement around the integration test around sorting. 

Have a look through the comments just to be sure. Happy to approve and let you get on with that - but feel free to put back to CR at all if you'd like a second view. 

RE: Repository maintenance then to begin, heading into the repository features and removing the Wiki and Projects seems a good idea. For Merge button features we tend to turn off `Allow merge commits`. Under Branches adding a protection rule for the `main` branch would be great too. I think you'll see branch protection rules for needing to pass CI as well as number of reviewers from the Artefactual organization. 

Your other points seem fine, for example, around a default output for auditmatica. As well as uploading to Pypi after a first round of CR. 

The README and IMPLEMENTATION.md look really awesome!

Really good work on all of this and bringing it together.
